cli/src/test/scala-2.12/coursier/cli/CliFetchIntegrationTest.scala

```diff
@@ -436,10 +436,77 @@ class CliFetchIntegrationTest extends FlatSpec with CliTestLib {
             )
           )
 
+          val node1: ReportNode = getReportFromJson(jsonFile)
+
+          val depNodes1: Seq[DepNode] = node1.dependencies
+            .filter(_.coord == "org.apache.commons:commons-compress:1.5")
+            .sortBy(fileNameLength)
+          assert(depNodes1.length == 1)
+
+          val urlInJsonFile1 = depNodes1.head.file.get
+          assert(urlInJsonFile1.contains(path))
+
+          // open jar and inspect contents
+          val fileContents1 = Source.fromFile(urlInJsonFile1).getLines.mkString
+          assert(fileContents1 == "tada")
+
+          testFile.delete()
+
+          Fetch.run(
+            fetchOpt,
+            RemainingArgs(
+              Seq(
+                "org.apache.commons:commons-compress:1.5,url=" + encodedUrl
+              ),
+              Seq()
+            )
+          )
+
+          val node2: ReportNode = getReportFromJson(jsonFile)
+
+          val depNodes2: Seq[DepNode] = node2.dependencies
+            .filter(_.coord == "org.apache.commons:commons-compress:1.5")
+            .sortBy(fileNameLength)
+          assert(depNodes2.length == 1)
+
+          val urlInJsonFile2 = depNodes2.head.file.get
+          assert(urlInJsonFile2.contains("coursier/cache") && urlInJsonFile2.contains("commons-compress-1.5.jar"))
+
+        }
+      }
+    }
+  }
+
+  /**
+   * Result:
+   * |└─ e:f:g
+   */
+  "local dep url" should "have coursier-fetch-test.jar but should fail when file is removed" in withFile() {
```

Comment:
This is a great test case!

On the other hand, there should be another case with `a:b:c,url=file://<path>` and `--cla`, so the first run put the file into cache, then remove the file locally, re-resolve should still work because it's in the cache. However it seems to fail currently:

```
$ fetch -t junit:junit:4.12,url=file%3A%2F%2F%2FUsers%2Fme%2F.m2%2Frepository%2Fjunit%2Fjunit%2F4.12%2Fjunit-4.12.jar%20  --no-default  --cla  --json-output-file x.out
  Result:
└─ junit:junit:4.12
/Users/me/Library/Caches/Coursier/v1/file/Users/me/.m2/repository/junit/junit/4.12/junit-4.12.jar

# move the repo, so the file://... isn't valid anymore
$ mv ~/.m2/repository/ ~/.m2/repository2

# this should still succeed by looking at the cache, instead of trying to look at file://...
$ fetch -t junit:junit:4.12,url=file%3A%2F%2F%2FUsers%2Fme%2F.m2%2Frepository%2Fjunit%2Fjunit%2F4.12%2Fjunit-4.12.jar%20  --no-default  --cla  --json-output-file x.out
```

Reply:
Alright, so looked into this a bit deeper and strangely, the fetch in the test above doesn't fail, but on the command line it does. I *think* (from what I can tell) the hack of using the [FallbackDependenciesRepository](https://github.com/coursier/coursier/blob/master/extra/src/main/scala/coursier/FallbackDependenciesRepository.scala) is what is causing the problem because it doesn't have caching behavior associated with it. @alexarchambault, could you confirm this?

Reply:
I've resolved this with my most recent commit, but I'm still stuck on trying to assert that a fetch failed without the artifact present. At the moment its causing the tests to exit, and asserting that an exception or error are thrown are not enough to catch the state. Additionally, I'd also love to get the test in `CacheFetchTests` working but I can't figure out how to import the `FallbackDependenciesRepository`. Hopefully @alexarchambault can help with those issues.